### PR TITLE
Add infrastructure for managing system indices

### PR DESCRIPTION
**Pull request by pugnascotia**
Part of #61656.

Add the necessary support for automatically creating and updating system
indices. This works by making it possible to create a system index
descriptor with all the information needed to manage the mappings,
settings and aliases.

Follow-up work will opt existing indices into this framework.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java:170]: Can we simplify this method a bit? We don't really share logic between non-system index creation and system index creation as far as I can see so I think we can test for the non system case and simply return the update request there?

File: server/src/main/java/org/elasticsearch/action/admin/indices/create/TransportCreateIndexAction.java
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/action/admin/indices/create/TransportCreateIndexAction.java:106]: Similar to the comment in `AutoCreateAction`, can we break up the system and non-system case since we don't really need to share anything between the two cases?

File: server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java:83]: I think we need to make an effort to wait for any pending updates to finish prior to launching more by tracking them between cluster state updates. We've had issues in the past where we overloaded master nodes with too many requests due to a pattern like this.

**Review by jaymode — CHANGES_REQUESTED** [review]
I took a brief look through the PR and left a few comments.

**Comment by pugnascotia** [comment]
Note to reviewers: I'm holding off creating follow-up PRs to port over existing systems indices until this is merged, so I'd appreciate any and all feedback. In particular, some guidance would be good for setting up a BWC test that exercises the mappings upgrade logic in `SystemIndexManager`.

File: server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/action/admin/indices/create/AutoCreateAction.java:196]: I think we should make this a debug log like the regular message

File: server/src/main/java/org/elasticsearch/action/admin/indices/settings/put/TransportUpdateSettingsAction.java
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/action/admin/indices/settings/put/TransportUpdateSettingsAction.java:100]: On backport, we will need to make this a deprecation warning

File: server/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/TransportPutMappingAction.java
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/TransportPutMappingAction.java:109]: On backport, we will need to make this a deprecation warning

File: server/src/main/java/org/elasticsearch/common/Strings.java
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/common/Strings.java:877]: Can you add a javadocs for this method to indicate what it does and what is thrown?

File: server/src/main/java/org/elasticsearch/indices/SystemIndexDescriptor.java
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/indices/SystemIndexDescriptor.java:242]: How about adding a private constructor so that all creations are from `SystemIndexDescriptor.builder()`?

File: server/src/main/java/org/elasticsearch/indices/SystemIndexDescriptor.java
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/indices/SystemIndexDescriptor.java:294]: Can all validation be moved into the constructor?

File: server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java:85]: Let's avoid extra work here by guarding this with a `if (upgradeInProgress.get() == false)` check

File: server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java:174]: this sets to false early. There can still be in flight items and isUpgradeInProgress would be incorrect. You may want to use something like a `GroupedActionListener`

**Review by jaymode — CHANGES_REQUESTED** [review]
I left some more feedback. I think this is close. Regarding testing overall, I think it might be hard without an index that uses this framework. So I think once we get this PR in and migrate one to use this then we can add tests that attempt to put mappings or settings that would trigger a violation.

File: server/src/main/java/org/elasticsearch/action/admin/indices/settings/put/TransportUpdateSettingsAction.java
Comment (pugnascotia) [reply server/src/main/java/org/elasticsearch/action/admin/indices/settings/put/TransportUpdateSettingsAction.java:100]: Can I clarify - in the backport, should this raise a deprecation warning but not fail the request?

File: server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java
Comment (pugnascotia) [reply server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java:174]: I'm not sure I follow. The `upgradeIndexMetadata` method is called once for each index that needs upgraded, and the first thing the method does is attempt to set `isUpgradeInProgress` to `true`. So only one index will be upgraded, and the rest need to wait until `isUpgradeInProgress` becomes false again. I wrote it that way so as to be cautious about adding load to the master.

Were you thinking that we'd dispatch all the mapping updates in one go, and use a `GroupedActionListener` to release the `isUpgradeInProgress` once they'd all finished?

File: server/src/main/java/org/elasticsearch/action/admin/indices/settings/put/TransportUpdateSettingsAction.java
Comment (jaymode) [reply server/src/main/java/org/elasticsearch/action/admin/indices/settings/put/TransportUpdateSettingsAction.java:100]: Correct on backport this should result in a deprecation warning and still allow the request to go through.

File: server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java
Comment (jaymode) [reply server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java:175]: > I'm not sure I follow. 

That's because I did not follow the code properly. I think it would be best to fire off all of the upgrades at once rather than sequentially since IIRC the code is computing everything that needs upgraded each time.

> Were you thinking that we'd dispatch all the mapping updates in one go, and use a GroupedActionListener to release the isUpgradeInProgress once they'd all finished?

Yes that is what I was thinking. I wanted to avoid resending each one every time there was a cluster state update if other cluster state updates happened prior to the mapping updates that we had previously sent.

Comment (pugnascotia) [reply server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java]: I've reworked the locking and upgrading to use a `GroupedActionListener`.

File: server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java:101]:
```suggestion
            } else {
                isUpgradeInProgress.set(false);
            }
```

**Review by jaymode — APPROVED** [review]
Left one comment. Otherwise LGTM. Thanks for the iterations!

File: server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java
Comment (pugnascotia) [reply server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java:103]: Good catch! 😅

**Pull request merged**
